syft/pkg/cataloger/cpe.go

```diff
@@ -323,15 +343,152 @@ func groupIDFromPomProperties(p pkg.Package) string {
+	if properties == nil {
+		return nil
+	}
+
+	if startsWithDomain(properties.GroupID) {
+		groupIDs = append(groupIDs, strings.TrimSpace(properties.GroupID))
+	}
+
+	// sometimes the publisher puts the group ID in the artifact ID field unintentionally
+	if startsWithDomain(properties.ArtifactID) && len(strings.Split(properties.ArtifactID, ".")) > 1 {
+		// there is a strong indication that the artifact ID is really a group ID
+		groupIDs = append(groupIDs, strings.TrimSpace(properties.ArtifactID))
+	}
+
+	return groupIDs
+}
+
+func groupIDsFromJavaManifest(manifest *pkg.JavaManifest) (groupIDs []string) {
+	if manifest == nil {
+		return nil
+	}
+	// attempt to get group-id-like info from the MANIFEST.MF "Automatic-Module-Name" and "Extension-Name" field.
+	// for more info see pkg:maven/commons-io/commons-io@2.8.0 within cloudbees/cloudbees-core-mm:2.263.4.2
+	// at /usr/share/jenkins/jenkins.war:WEB-INF/plugins/analysis-model-api.hpi:WEB-INF/lib/commons-io-2.8.0.jar
+	// as well as the ant package from cloudbees/cloudbees-core-mm:2.277.2.4-ra.
+	for name, value := range manifest.Main {
+		value = strings.TrimSpace(value)
+		switch name {
+		case "Extension-Name", "Automatic-Module-Name":
+			if startsWithDomain(value) {
+				groupIDs = append(groupIDs, value)
+			}
+		}
+	}
+	for _, section := range manifest.NamedSections {
+		for name, value := range section {
+			value = strings.TrimSpace(value)
+			switch name {
+			case "Extension-Name", "Automatic-Module-Name":
+				if startsWithDomain(value) {
+					groupIDs = append(groupIDs, value)
+				}
+			}
+		}
+	}
+	return groupIDs
+}
+
+func startsWithDomain(value string) bool {
+	return internal.HasAnyOfPrefixes(value, domains...)
+}
+
+func generateAllSubSelections(fields []string) (results []string) {
+	for _, field := range fields {
+		results = append(results, generateSubSelections(field)...)
+	}
+	return results
+}
+
+// generateSubSelections attempts to split a field by hyphens and underscores and return a list of sensible sub-selections
+// that can be used as product or vendor candidates. E.g. jenkins-ci-tools -> [jenkins-ci-tools, jenkins-ci, jenkins].
+func generateSubSelections(field string) (results []string) {
+	scanner := bufio.NewScanner(strings.NewReader(field))
+	scanner.Split(scanByHyphenOrUnderscore)
+	var lastToken uint8
+	for scanner.Scan() {
+		rawCandidate := scanner.Text()
+		if len(rawCandidate) == 0 {
+			break
+		}
+
+		candidate := strings.TrimFunc(rawCandidate, trimHyphenOrUnderscore)
+
+		// capture the result (if there is content)
+		if len(candidate) > 0 {
+			if len(results) > 0 {
+				results = append(results, results[len(results)-1]+string(lastToken)+candidate)
+			} else {
+				results = append(results, candidate)
+			}
+		}
+
+		// keep track of the trailing separator for the next loop
+		lastToken = rawCandidate[len(rawCandidate)-1]
+	}
+	return results
+}
+
+func trimHyphenOrUnderscore(r rune) bool {
```

Comment:
It'd be nice to have a doc comment that explains why this function exists and what it does. (I did figure this out through looking around, but it'd be nice to have the answer more immediately if possible.)